lib/virt_autotest/utils.pm

```diff
@@ -42,62 +42,89 @@ sub trim {
-        assert_script_run("kill -9 $_libvirtd_pid");
-        assert_script_run("$_libvirtd_cmd");
-    }
-    else {
-        systemctl("restart libvirtd", timeout => 180);
-    }
-    record_info("Libvirtd has been restarted!");
-    save_screenshot;
-}
-
 #return 1 if test is expected to run on XEN hypervisor
 sub is_xen_host {
     return get_var("XEN") || check_var("SYSTEM_ROLE", "xen") || check_var("HOST_HYPERVISOR", "xen") || check_var("REGRESSION", "xen-hypervisor");
 }
 
+# Usage: check_modular_libvirt_daemons([daemon1_name daemon2_name ...]). For example:
+# to specify daemons which will be checked: check_modular_libvirt_daemons(qemu storage ...)
+# to check all required modular daemons without any daemons passed
+sub check_modular_libvirt_daemons {
+    my @daemons = @_;
+
+    if (!@daemons) {
+        @daemons = qw(network nodedev nwfilter secret storage proxy log lock);
+        (is_xen_host) ? push @daemons, 'xen' : push @daemons, 'qemu';
+    }
+
+    foreach my $daemon (@daemons) {
+        systemctl("status virt${daemon}d.service");
+        if (($daemon == 'lock') || ($daemon == 'log')) {
+            systemctl("status virt${daemon}d\{,-admin\}.socket");
+        } else {
+            systemctl("status virt${daemon}d\{,-ro,-admin\}.socket");
+        }
+    }
+
+    record_info("Modular Libvirt daemons checked", join(' ', @daemons));
+    save_screenshot;
+}
+
 # Usage: restart_modular_libvirt_daemons([daemon1_name daemon2_name ...]). For example:
 # to specify daemons which will be restarted: restart_modular_libvirt_daemons(virtqemud virtstoraged ...)
 # to restart all modular daemons without any daemons passed
 sub restart_modular_libvirt_daemons {
-    my @daemons;
+    my @daemons = @_;
 
-    if (@_ == 0) {
-        if (is_xen_host) {
-            @daemons = qw(virtxend virtstoraged virtnetworkd virtnodedevd virtsecretd virtproxyd virtnwfilterd);
-        } else {
-            @daemons = qw(virtqemud virtstoraged virtnetworkd virtnodedevd virtsecretd virtproxyd virtnwfilterd);
-        }
-    } else {
-        @daemons = @_;
+    if (!@daemons) {
+        @daemons = qw(network nodedev nwfilter secret storage proxy log lock);
+        (is_xen_host) ? push @daemons, 'xen' : push @daemons, 'qemu';
     }
 
     if (is_alp) {
         record_soft_failure("Restarting modular libvirt daemons has not been implemented in ALP. See poo#129086");
     } else {
         # Restart the sockets first
         foreach my $daemon (@daemons) {
-            systemctl("restart $daemon\{,-ro,-admin\}.socket");
+            if (($daemon == 'lock') || ($daemon == 'log')) {
+                systemctl("restart virt${daemon}d\{,-admin\}.socket");
+            } else {
+                systemctl("restart virt${daemon}d\{,-ro,-admin\}.socket");
+            }
         }
 
         # Introduce idle time here (e.g., sleep 5) if necessary
-        sleep 5;
+        script_retry("virsh list", delay => 10, retry => 1, die => 0) for (0 .. 4);
```

Comment:
Is executing "virsh list" in the "middle" of a complex restart good? I'd vote for 'sleep 5',which is simpler.

Reply:
Compare with ```sleep```, I think using following is better and also safe:
```
host # systemctl list-units virt*
  UNIT                  LOAD   ACTIVE SUB       DESCRIPTION                        
  virtlogd.service      loaded active running   Virtual machine log manager        
  virtlockd.socket      loaded active listening Virtual machine lock manager socket
  virtlogd-admin.socket loaded active running   Virtual machine log manager socket
  virtlogd.socket       loaded active running   Virtual machine log manager socket

LOAD   = Reflects whether the unit definition was properly loaded.
ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
SUB    = The low-level unit activation state, values depend on unit type.
4 loaded units listed. Pass --all to see loaded but inactive units, too.
To show all installed unit files use 'systemctl list-unit-files'.

host # systemctl list-sockets virt*
LISTEN                           UNIT                  ACTIVATES        
/run/libvirt/virtlockd-sock      virtlockd.socket      virtlockd.service
/run/libvirt/virtlogd-admin-sock virtlogd-admin.socket virtlogd.service
/run/libvirt/virtlogd-sock       virtlogd.socket       virtlogd.service
```

Reply:
@guoxuguang  Hi Leon, based on your validation jobs, the above script_retry will take 3*10*5=150 seconds, and it only executes commands, no assert at all. Would you please explain the necessity/purpose for it? To me, it does not quite necessary or need that long, better improve it from overall efficiency regard. 
(1)
If I understand the purpose correctly that with `virsh list ` check, we'd claim if all is ok after restart. Then I'd recommend to use 

 `sleep 5`  here and  
move script_retry to the line after all services are restarted(line 104) with assert, sth like  `script_retry("virsh list --all", delay => 3, retry => 3, die => 1)` 
(2) above wayne's suggestion is also ok, but also better to add assert and move it after line 104.

BTW, let's make this clear before rerunning last round of validation jobs to avoid wasting effort run again. Thanks.

Reply:
I am curious about the purpose of `script_retry("virsh list")` either. It took 27s in my TW run, https://openqa.opensuse.org/tests/3597532/logfile?filename=autoinst-log.txt.

Anyway, I would suggest move enabling debugging log to host autoyast installation to skip service restart when module libivrt goes to SLE16sp6 later.

Reply:
I need to explain here that it was originally intended to replace sleep. But now it seems that this replacement method has caused a lot of misunderstandings. So, I'd vote for "Sleep 5," which is simple too.